### ISSUE TITLE
Fetch all existing providers

### ISSUE THREAD

**Issue by Azbesciak** [angular/angular#43304]
### Which @angular/* package(s) are relevant/releated to the feature request?

common

### Description

Hello,
Let us assume that we have 3 levels of the same provider provision - each on the component `providers` scope.
This token is provided with `multi: true`.
The last one can access his own provision and if specified in another constructor field `@SkipSelf()`, also his parent. But, unfortunately, there is no option to access provision from the first level.
So like bellow
- component A provides instance A.1 - it can access only this A.1
- component B provides instance B.1 - it can access his B.1 and A.1 of his parent
- component C provides instance C.1 - it can access his C.1 and his parent B.1, but cannot A.1

You can see the live example here: https://codesandbox.io/s/peaceful-hamilton-0nejz?file=/src/app/deep-test/deep-test.component.ts

From C I would like also to access the A.1 instance, and if I would have component Z - also A.1, B.1, E.1, R.1 and so on.

### Proposed solution

Provide an Injector flag/annotation like "@All" or "@Deep" which try to find all provisions of the given provider. 

### Alternatives considered

For now even manually by checking with `Injector.get(..., InjectorFlags.SkipSelf)` it is impossible or with recursion by fetching the injector of the current injector (it is endless and I did not find any way to stop it - each instance is different),

**Comment by petebacondarwin** [comment]
I believe this is really a duplicate of https://github.com/angular/angular/issues/18894, where we really have a limitation in our DI system of how we handle multiproviders that are overridden in components.

**Comment by Azbesciak** [comment]
Possible, also connected with https://github.com/angular/angular/pull/41592 and https://github.com/angular/angular/pull/39464 I think

**Comment by JoostK** [comment]
We have discussed this request and will no go forward with adding such feature, as this is already possible with today's APIs.

Since #41592 landed it is now possible to use `InjectFlags.SkipSelf` to obtain the parent injector, allowing to iterate the injector chain manually.

As for `multi: true` providers, you can already merge these today by having an accumulator token using a factory function:

```ts
const TOKEN = new InjectionToken<SomeType[]>('some description');
const TOKEN_ACCUMULATOR = new InjectionToken<SomeType[]>('some description accumulator', {
  providedIn: 'any',
  factory: () => {
    const self = inject(TOKEN);
    const parent = inject(TOKEN, InjectFlags.SkipSelf | InjectFlags.Optional);
    if (parent === null) {
      return self;
    }
    return [...self, ...parent];
  },
});
```

Or alternatively using an explicit provider for the accumulator token:

```ts
const accumulatorProvider = {
  provide: TOKEN_ACCUMULATOR,
  useFactory: (self: SomeType[], parent: SomeType[]|null) => {
    if (parent === null) {
      return self;
    }
    return [...self, ...parent];
  },
  deps: [TOKEN, [new SkipSelf(), new Optional(), TOKEN]],
}
```

This way, consumers provide things using the `TOKEN` token, whereas `TOKEN_ACCUMULATOR` is used internally to inject to values.

**Comment by Azbesciak** [comment]
@JoostK , I appreciate that it is possible, but now imagine that in my project I actually use only use dependencies like this (I have a customizable application that also acts as a plug-in in another app) - so, in this way, I should duplicate all my tokens. Not to mention that similar code would be duplicated in many projects, whereas a more 'humanlike' approach would be to place this in the framework because the problem also comes from its imperfection.

IMHO it would be enough to provide another annotation, like `@All`, or to configure `@SkipSelf` to allow it to look further, not only at the 2nd dependencies tree-level...

BTW, if we are talking about dependencies level, tell me where from I would know where to place another factory?
By now I see it through the mist, but I remember that I tried to fetch it manually, from the injector (without factory), and it was endless - fetching the injector caused the new injector creation.

**Comment by JoostK** [comment]
> so, in this way, I should duplicate all my tokens.

If the approach of duplicating tokens isn't desirable in your case, an alternative would be to have a function that iterates the injector chain manually.

> Not to mention that similar code would be duplicated in many projects, whereas a more 'humanlike' approach would be to place this in the framework because the problem also comes from its imperfection.

I feel quite the opposite actually: the necessary APIs are available and allow composing behavior as desired. There is no single approach to merging tokens from multiple levels of an injector chain. Some hypothetical design considerations: do you prepend or append? Prioritize somehow? Do you filter out duplicates? I believe that the answer to these questions is "it depends on a specific use-case" and so providing it from the framework would not suit everyone while only increasing the API surface of DI, introducing additional concepts to document and for people to learn.

> BTW, if we are talking about dependencies level, tell me where from I would know where to place another factory?
By now I see it through the mist, but I remember that I tried to fetch it manually, from the injector (without factory), and it was endless - fetching the injector caused the new injector creation.

That was likely due to #41592, which was recently resolved.

**Comment by Azbesciak** [comment]
>do you prepend or append? Prioritize somehow
By now I believe that it depends on the order provided in the `providers` section, and I suppose this should be done in the same way this way. Also, given dependency interface can offer something like `priority`, and be sorted in consumers by their usage needs, so IMO this is not the case.

> That was likely due to #41592, which was recently resolved.
Ok, did not test it yet, but does it mean that getting from the injector to its parent injector won't cause generating a new injector each time, like in my original problem? As I wrote I tried to fetch all tokens of a given type manually, but because of this, there was no end of the tree (no root).

**Comment by JoostK** [comment]
> Ok, did not test it yet, but does it mean that getting from the injector to its parent injector won't cause generating a new injector each time, like in my original problem? As I wrote I tried to fetch all tokens of a given type manually, but because of this, there was no end of the tree (no 'stable' root).

It used to be that programmatically providing `InjectFlags.SkipSelf` would be ignored by node injectors, which was indeed a problem for requesting the parent injector. This was fixed in #41592, so now this should work fine.
